File: sub_pub/flows/fan.py
"""Fan flow: one source -> many destinations"""
import logging
import threading
import json

from sub_pub.flows.base import Flow
from sub_pub.core.interfaces import MessageSource, MessagePublisher
from sub_pub.core.message import Message

logger = logging.getLogger(__name__)


class FanFlow(Flow):
    """Fan flow: read from one source, publish to many destinations based on message content"""
    
    def __init__(
        self,
        source: MessageSource,
        source_topic: str,
        destination: MessagePublisher,
        destination_resolver: dict,  # {'type': 'header'/'payload_key', 'key': 'topic_name'}
        **kwargs
    ):
        super().__init__(**kwargs)
        self.source = source
        self.source_topic = source_topic
        self.destination = destination
        self.destination_resolver = destination_resolver
        
    def _resolve_destination_topic(self, message: Message) -> str:
        """Resolve the destination topic from the message"""
        resolver_type = self.destination_resolver['type']
        key = self.destination_resolver['key']
        
        if resolver_type == 'header':
            return message.get_header(key, 'default')
        elif resolver_type == 'payload_key':
            try:
                payload = json.loads(message.payload.decode('utf-8'))
                return payload.get(key, 'default')
            except Exception as e:
                logger.warning(f"Failed to parse payload for destination resolution: {e}")
                return 'default'
        else:
            logger.warning(f"Unknown resolver type: {resolver_type}")
            return 'default'
    
    def _consume_messages(self) -> None:
        """Consume messages from source"""
        try:
            for message in self.source.consume():
                if not self._running:
                    break
                
                try:
                    # Record source metrics
                    self.metrics.record_source_message(message.topic, message.size())
                    
                    # Resolve destination topic
                    dest_topic = self._resolve_destination_topic(message)
                    
                    # Apply back-pressure if needed
                    if self._check_back_pressure(self.domain_queue):
                        logger.debug("Back-pressure detected on domain queue, waiting...")
                        self._wait_for_back_pressure_release(self.domain_queue)
                    
                    # Add to domain queue
                    self.domain_queue.put((message, dest_topic))
                    
                    # Commit the message
                    self.source.commit(message)
                    
                except Exception as e:
                    logger.error(f"Error processing message: {e}", exc_info=True)
                    self.metrics.record_source_error(message.topic)
        except Exception as e:
            logger.error(f"Error consuming messages: {e}", exc_info=True)
    
    def run(self) -> None:
        """Run the fan flow"""
        self._running = True
        
        # Connect source and destination
        self.source.connect()
        self.source.subscribe([self.source_topic])
        self.destination.connect()
        
        # Start domain workers
        domain_threads = []
        for i in range(self.thread_pool_config.max_workers):
            t = threading.Thread(
                target=self._domain_worker,
                name=f"domain-worker-{i}",
                daemon=True
            )
            t.start()
            domain_threads.append(t)
        
        # Start publish workers
        publish_threads = []
        for i in range(self.thread_pool_config.max_workers):
            t = threading.Thread(
                target=lambda: self._publish_worker(self.destination),
                name=f"publish-worker-{i}",
                daemon=True
            )
            t.start()
            publish_threads.append(t)
        
        # Start consumer thread
        consumer_thread = threading.Thread(
            target=self._consume_messages,
            name="consumer",
            daemon=True
        )
        consumer_thread.start()
        
        logger.info(f"Fan flow started from {self.source_topic}")
        
        # Wait for shutdown signal
        self._shutdown_event.wait()
        
        # Cleanup
        consumer_thread.join(timeout=5.0)
        for thread in domain_threads + publish_threads:
            thread.join(timeout=5.0)
        
        # Flush and close
        self.destination.flush()
        self.destination.close()
        self.source.close()
        
        logger.info("Fan flow stopped")

File: rust/src/flows/fan.rs
use std::sync::Arc;

use serde_json::Value as JsonValue;
use tracing::{error, info, warn};

use crate::config::{DestinationResolver, FanConfig};
use crate::domain::MessageProcessor;
use crate::error::SubPubError;
use crate::metrics::MetricsCollector;

use super::base::{apply_back_pressure, build_pipeline, spawn_domain_worker, PipelineItem};

/// Fan flow: one source → many destinations (content-based routing).
///
/// The destination topic for each message is resolved either from a message
/// header or from a field in the JSON payload, according to
/// [`DestinationResolver`].  Messages that cannot be resolved fall back to a
/// topic named `"default"`.
pub struct FanFlow {
    cfg: FanConfig,
    processor: Arc<dyn MessageProcessor>,
    metrics: Arc<MetricsCollector>,
}

impl FanFlow {
    pub fn new(
        cfg: FanConfig,
        processor: Arc<dyn MessageProcessor>,
        metrics: Arc<MetricsCollector>,
    ) -> Self {
        Self {
            cfg,
            processor,
            metrics,
        }
    }

    fn resolve_destination(&self, item: &PipelineItem) -> String {
        match &self.cfg.destination_resolver {
            DestinationResolver::Header { header_key } => {
                item.message.get_header(header_key, "default").to_string()
            }
            DestinationResolver::PayloadKey { payload_key } => {
                let parsed: Result<JsonValue, _> =
                    serde_json::from_slice(&item.message.payload);
                match parsed {
                    Ok(JsonValue::Object(map)) => map
                        .get(payload_key)
                        .and_then(|v| v.as_str())
                        .unwrap_or("default")
                        .to_string(),
                    _ => "default".to_string(),
                }
            }
        }
    }

    /// Run the flow until `shutdown_rx` fires or a fatal error occurs.
    pub async fn run(
        self,
        shutdown_rx: tokio::sync::watch::Receiver<bool>,
    ) -> Result<(), SubPubError> {
        use crate::adapters::{create_publisher, create_source};

        let queue_size = self.cfg.thread_pool.queue_size;
        let (domain_tx, publish_tx, domain_rx, mut publish_rx) =
            build_pipeline(queue_size, queue_size);

        // Spawn domain-processing worker.
        let _domain_handle = spawn_domain_worker(
            domain_rx,
            publish_tx,
            Arc::clone(&self.processor),
            Arc::clone(&self.metrics),
        );

        // Spawn publish worker.
        let mut publisher = create_publisher(&self.cfg.destination)?;
        publisher.connect().await?;

        let metrics_pub = Arc::clone(&self.metrics);
        let _publish_handle = tokio::spawn(async move {
            while let Some(item) = publish_rx.recv().await {
                let topic = &item.destination_topic;
                let size = item.message.size();
                match publisher.publish(&item.message, topic).await {
                    Ok(()) => metrics_pub.record_destination_message(topic, size),
                    Err(e) => {
                        error!(%e, %topic, "publish error");
                        metrics_pub.record_destination_error(topic);
                    }
                }
            }
            let _ = publisher.flush().await;
            let _ = publisher.close().await;
        });

        // Connect source.
        let mut source = create_source(&self.cfg.source)?;
        source.connect().await?;
        source
            .subscribe(&[&self.cfg.source_topic])
            .await?;

        info!("FanFlow started; source topic: {}", self.cfg.source_topic);

        loop {
            if *shutdown_rx.borrow() {
                break;
            }

            apply_back_pressure(&self.cfg.back_pressure, &domain_tx).await;

            match source.consume().await {
                Ok(Some(msg)) => {
                    let src_topic = msg.topic.clone();
                    let size = msg.size();
                    self.metrics.record_source_message(&src_topic, size);

                    // Resolve destination topic before moving `msg` into
                    // PipelineItem (we need message headers / payload).
                    let item = PipelineItem {
                        destination_topic: String::new(), // placeholder
                        message: msg,
                    };
                    let dest_topic = self.resolve_destination(&item);

                    if domain_tx
                        .send(PipelineItem {
                            destination_topic: dest_topic,
                            ..item
                        })
                        .await
                        .is_err()
                    {
                        warn!("domain channel closed; stopping fan consumer");
                        break;
                    }
                }
                Ok(None) => {
                    info!("source exhausted; stopping flow");
                    break;
                }
                Err(e) => {
                    error!(%e, "fan source consume error");
                    self.metrics.record_source_error(&self.cfg.source_topic);
                }
            }

            if shutdown_rx.has_changed().unwrap_or(false) && *shutdown_rx.borrow() {
                break;
            }
        }

        drop(domain_tx);
        let _ = source.close().await;
        info!("FanFlow stopped");
        Ok(())
    }
}
